cwlVersion: v1.2
class: Workflow
label: NULL Model SVA
$namespaces:
  sbg: https://sevenbridges.com

requirements:
- class: SubworkflowFeatureRequirement
- class: InlineJavascriptRequirement
- class: StepInputExpressionRequirement

inputs:
- id: input_gds_files
  label: GDS files
  doc: |-
    GDS files with genotype data for variants to be tested for association. If multiple files are selected, they will be run in parallel. Files separated by chromosome are expected to have ‘chr##’ strings indicating chromosome number, where ‘##’ can be (1-24, X, Y). Output files for each chromosome will include the corresponding chromosome number.
  type: File[]
  sbg:fileTypes: GDS
  sbg:x: 445.3125
  sbg:y: 111.921875
- id: phenotype_file
  label: Phenotype file
  doc: |-
    RData file with an AnnotatedDataFrame of phenotypes and covariates. Sample identifiers must be in column named “sample.id”.
  type: File
  sbg:fileTypes: RDATA
  sbg:x: -15
  sbg:y: -89.5390625

outputs:
- id: html_reports
  label: HTML Reports
  doc: HTML Reports generated by the tool.
  type: File[]?
  outputSource:
  - null_model/html_reports
  sbg:fileTypes: html
  sbg:x: 317.2303161621094
  sbg:y: 245.921875
- id: assoc_plots
  label: Association test plots
  doc: QQ and Manhattan Plots of p-values in association test results.
  type: File[]?
  outputSource:
  - single_variant_association_testing/assoc_plots
  sbg:fileTypes: PNG
  sbg:x: 797.121826171875
  sbg:y: -208
- id: assoc_combined
  label: Association test results
  doc: |-
    RData file with data.frame of association test results (test statistic, p-value, etc.) See the documentation of the GENESIS R package for detailed description of output.
  type: File[]?
  outputSource:
  - single_variant_association_testing/assoc_combined
  sbg:fileTypes: RDATA
  sbg:x: 807.121826171875
  sbg:y: 13.921875

steps:
- id: null_model
  label: Null Model
  in:
  - id: phenotype_file
    source: phenotype_file
  run: null-model-sva-unpacked.cwl.steps/null_model.cwl
  out:
  - id: null_model_phenotypes
  - id: rmd_files
  - id: html_reports
  - id: null_model_file
  sbg:x: 157.3125
  sbg:y: 0
- id: single_variant_association_testing
  label: GENESIS Single Variant Association Testing
  in:
  - id: phenotype_file
    source: null_model/null_model_phenotypes
  - id: null_model_file
    source: null_model/null_model_file
  - id: input_gds_files
    source:
    - input_gds_files
  run: null-model-sva-unpacked.cwl.steps/single_variant_association_testing.cwl
  out:
  - id: assoc_combined
  - id: assoc_plots
  sbg:x: 551.2303466796875
  sbg:y: -87
sbg:appVersion:
- v1.2
- v1.1
sbg:content_hash: afa7bd3fed79e1725d9eac4d2ac4c0967aa367c8274ad62c681807d7ec820e545
sbg:contributors:
- dave
sbg:createdBy: dave
sbg:createdOn: 1621269323
sbg:id: dave/build-demo-gwas-workflows/null-model-sva/1
sbg:image_url: |-
  https://platform.sb.biodatacatalyst.nhlbi.nih.gov/ns/brood/images/dave/build-demo-gwas-workflows/null-model-sva/1.png
sbg:latestRevision: 1
sbg:modifiedBy: dave
sbg:modifiedOn: 1621269421
sbg:original_source: |-
  https://api.sb.biodatacatalyst.nhlbi.nih.gov/v2/apps/dave/build-demo-gwas-workflows/null-model-sva/1/raw/
sbg:project: dave/build-demo-gwas-workflows
sbg:projectName: 'BUILD: Demo GWAS Workflows'
sbg:publisher: sbg
sbg:revision: 1
sbg:revisionNotes: ''
sbg:revisionsInfo:
- sbg:modifiedBy: dave
  sbg:modifiedOn: 1621269323
  sbg:revision: 0
  sbg:revisionNotes:
- sbg:modifiedBy: dave
  sbg:modifiedOn: 1621269421
  sbg:revision: 1
  sbg:revisionNotes: ''
sbg:sbgMaintained: false
sbg:validationErrors:
- |-
  Missing one or more mandatory run inputs in step inputs: ['#null_model.outcome', '#null_model.family']
